dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+// stack is made explicit through the `path` variable.
+//
+// It exposes simple traversal operations like `Iterate` and  `Search`
+// that are implemented through a series of simple move methods (`down`,
+// `up`, `Right`) that modify its `path`. The `down` move method is the
+// analogous of the recursive call and the one in charge of visiting
+// the node (through the `VisitHandler`) and performing some user-defined
+// logic.
+//
+// The position of the `Traversal` is defined through a `path` (slice)
+// of nodes that form a parent-child relationship from the root to the
+// current node being visited (at a depth marked by the `level`). Each
+// of those nodes keep track of the link index its child in the path
+// belongs to in order for the traversal algorithms to know not only
+// what it the current nodes but which nodes (children) have already
+// been visited (assuming always a left to right, zero to `len(Links())`,
+// visiting order).
+//
+// TODO: Add a logic similar to `precalcNextBuf` to request several child
+// nodes in advanced as node promises.
+//
+// TODO: Revisit the name. `Traverser`? (is "traverser" the correct noun?),
+// `Iterator` (this would put to much emphasis on iterate whereas other
+// traversal operations like search are supported), `Topology`?
+//
+// TODO: Encapsulate the position (`path`, `level`, `childIndex`)? It's
+// a big part of the structure, and it should only be one of it, so it
+// would just create a proxy for every move call, e.g., `dt.pos.down()`.
+//
+// TODO: Consider adding methods that would retrieve an attribute indexed
+// by the current `level` to avoid including much indexing like
+// `dt.path[dt.level]` in the code. Maybe another more strong refactoring
+// that would allow to think of the current node in the `path` without
+// concerning at what `level` it's in.
+type Traversal struct {
+
+	// Each member of the slice is the parent of the following member, from
+	// the root to the current node, up until the position marked by `level`. The
+	// slice may contain more elements past that point but they should be
+	// ignored, every time the `level` increases the corresponding child node
+	// is inserted (the slice is not truncated to leverage the already allocated
+	// space). It should *always* have a length bigger than zero with the root
+	// of the DAG at the first position (empty DAGs are not valid).
+	path []ipld.Node
+
+	// This slice has the index of the child each parent in `path` is pointing
+	// to. Its current valid length is determined by `level`. The index in the
+	// parent can be set past all of its child nodes (having a value equal to
+	// `len(Links())`) to signal it has visited (or skipped) all of them. A
+	// leaf node with no children and its index in zero would also comply with
+	// this format.
+	childIndex []uint
+	// TODO: Revisit name, `childPosition`? Do not use the *link* term, put
+	// emphasis in the child (the *what*, not the *how* we get it).
```

Comment:
I think this name made plenty of sense for me. The only thing clearer that comes to mind is `currentChildIndex`.